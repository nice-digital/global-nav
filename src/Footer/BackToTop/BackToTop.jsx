import React from "react";
import ChevronUp from "@nice-digital/icons/lib/ChevronUp";
import { Container } from "@nice-digital/nds-container";
import PropTypes from "prop-types";

import styles from "./BackToTop.module.scss";

export const BackToTop = function ({
	scrollTargetId: scrollTo = "content-start",
}) {
	return (
		<div className={styles.wrapper}>
			<nav aria-labelledby="back-to-top-link" className={styles.nav}>
				<a
					className={styles.anchor}
					id="back-to-top-link"
					href={`#${scrollTo}`}
				>
					<Container>
						<ChevronUp /> Back to top
					</Container>
				</a>
			</nav>
		</div>
	);
};

BackToTop.propTypes = {
	show: PropTypes.boolean,
	scrollTargetId: PropTypes.string,
};
