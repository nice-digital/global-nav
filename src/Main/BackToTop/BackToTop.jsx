import React from "react";
import ChevronUp from "@nice-digital/icons/lib/ChevronUp";
import { Container } from "@nice-digital/nds-container";

import styles from "./BackToTop.module.scss";

export const BackToTop = function () {
	return (
		<div className={styles.wrapper}>
			<nav aria-labelledby="back-to-top-link" className={styles.nav}>
				<a className={styles.anchor} id="back-to-top-link" href="#top">
					<Container>
						<ChevronUp /> Back to top
					</Container>
				</a>
			</nav>
		</div>
	);
};
