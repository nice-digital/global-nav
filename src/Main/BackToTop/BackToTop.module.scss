@import '~@nice-digital/nds-core/scss/core';

$back-to-top-height: rem(
  $nds-spacing-medium * 3
); // Line height + bottom and top padding

.wrapper {
  margin-bottom: rem($nds-spacing-large) * -1;
  margin-top: rem($nds-spacing-x-large);

  @media print {
    display: none;
  }
}

.nav {
  background: $nds-colour-nice-grey-3;
  border-top: 1px solid $nds-colour-border;
  width: 100%;
  z-index: 1;
}

.anchor {
  display: block;
  line-height: rem($nds-spacing-medium);
  padding: rem($nds-spacing-medium) 0;
  text-decoration: underline;

  &:visited {
    color: $nds-colour-link;
  }

  svg {
    margin-right: 0.5rem;
    transition: all 0.25s ease;
  }

  &:hover {
    background: $nds-colour-body;

    svg {
      transform: translateY(-#{rem($nds-spacing-xx-small)});
    }
  }

  .container {
    display: flex;
  }
}

@supports (position: sticky) {
  .wrapper {
    bottom: 0;
    left: 0;
    margin: 0;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: min(100vh, calc(100% - #{$back-to-top-height}));
    width: 100%;
    z-index: 99;
  }

  .nav {
    background: $nds-colour-nice-grey-3;
    border-top: 1px solid $nds-colour-border;
    pointer-events: all;
    position: sticky;
    top: calc(100vh - #{$back-to-top-height});
    width: 100%;
    z-index: 1;
  }
}
