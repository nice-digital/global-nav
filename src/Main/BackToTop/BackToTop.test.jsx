import { BackToTop } from "./BackToTop";
import { render, createEvent, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("BackToTop", () => {
	const scrollIntoViewMock = jest.fn();

	let top;

	beforeEach(() => {
		top = document.createElement("div");
		top.id = "top";
		top.scrollIntoView = scrollIntoViewMock;
		document.body.appendChild(top);
	});

	afterEach(() => {
		document.body.innerHTML = "";
		scrollIntoViewMock.mockReset();
	});

	it("Matches snapshot", () => {
		const { container } = render(<BackToTop />);
		expect(container).toMatchSnapshot();
	});

	it("Renders a link with an accessible name", () => {
		const { getByRole } = render(<BackToTop />);
		expect(getByRole("link")).toHaveAccessibleName("Back to top");
	});

	it("Renders a link with the default hash when no scroll target id prop provided", () => {
		const { getByRole } = render(<BackToTop />);
		expect(getByRole("link")).toHaveAttribute("href", "#top");
	});

	it("Prevents default and moves focus to the back to top link target on click", () => {
		const { getByRole } = render(<BackToTop />),
			link = getByRole("link"),
			clickEvent = createEvent.click(link);

		fireEvent(link, clickEvent);

		expect(clickEvent.defaultPrevented).toBe(true);
		expect(top).toHaveAttribute("tabindex", "-1");
		expect(top).toHaveFocus();
	});

	it("Scrolls element with id = back to top link target into view on back to top click", async () => {
		const { getByRole } = render(<BackToTop />),
			link = getByRole("link"),
			user = userEvent.setup();

		await user.click(link);

		expect(scrollIntoViewMock).toHaveBeenCalledTimes(1);
	});
});
