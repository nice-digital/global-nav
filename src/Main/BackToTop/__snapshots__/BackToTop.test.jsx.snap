// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop Matches snapshot 1`] = `
<div
  className="wrapper"
>
  <nav
    aria-labelledby="back-to-top-link"
    className="nav"
  >
    <a
      className="anchor"
      href="#top"
      id="back-to-top-link"
      onClick={[Function]}
    >
      <Container
        className="container"
      >
        <SvgChevronUp />
         Back to top
      </Container>
    </a>
  </nav>
</div>
`;
